File: Example/FWFramework/Classes/Module/Test/Module/TestBadgeController.m
//
//  TestBadgeController.m
//  FWFramework_Example
//
//  Created by wuyong on 2022/8/25.
//  Copyright © 2022 CocoaPods. All rights reserved.
//

#import "TestBadgeController.h"
#import "AppSwift.h"
@import FWFramework;

@interface TestBadgeController () <FWViewController>

@end

@implementation TestBadgeController

- (void)didInitialize
{
    self.hidesBottomBarWhenPushed = NO;
    self.fw_extendedLayoutEdge = UIRectEdgeBottom;
}

- (void)viewDidLoad
{
    [super viewDidLoad];
    
    self.fw_tabBarHidden = NO;
    
    [self fw_setLeftBarItem:FWIcon.backImage target:self action:@selector(onClose)];
    FWBadgeView *badgeView = [[FWBadgeView alloc] initWithBadgeStyle:FWBadgeStyleDot];
    [self.navigationItem.leftBarButtonItem fw_showBadgeView:badgeView badgeValue:nil];
    
    UIBarButtonItem *rightItem = [UIBarButtonItem fw_itemWithObject:FWIcon.backImage target:self action:@selector(onClick:)];
    badgeView = [[FWBadgeView alloc] initWithBadgeStyle:FWBadgeStyleSmall];
    [rightItem fw_showBadgeView:badgeView badgeValue:@"1"];
    
    UIButton *customView = [[UIButton alloc] initWithFrame:CGRectMake(0, 0, 20, 20)];
    customView.backgroundColor = [AppTheme textColor];
    UIBarButtonItem *customItem = [UIBarButtonItem fw_itemWithObject:customView target:self action:@selector(onClick:)];
    badgeView = [[FWBadgeView alloc] initWithBadgeStyle:FWBadgeStyleSmall];
    [customItem fw_showBadgeView:badgeView badgeValue:@"1"];
    self.navigationItem.rightBarButtonItems = @[rightItem, customItem];
}

- (void)viewWillAppear:(BOOL)animated
{
    [super viewWillAppear:animated];
    
    FWBadgeView *badgeView = [[FWBadgeView alloc] initWithBadgeStyle:FWBadgeStyleDot];
    [self.tabBarController.tabBar.items[0] fw_showBadgeView:badgeView badgeValue:nil];
    badgeView = [[FWBadgeView alloc] initWithBadgeStyle:FWBadgeStyleSmall];
    [self.tabBarController.tabBar.items[1] fw_showBadgeView:badgeView badgeValue:@"99"];
}

- (void)viewWillDisappear:(BOOL)animated
{
    [super viewWillDisappear:animated];
    
    [self.tabBarController.tabBar.items[0] fw_hideBadgeView];
    [self.tabBarController.tabBar.items[1] fw_hideBadgeView];
}

- (void)setupSubviews
{
    UIView *view = [[UIView alloc] initWithFrame:CGRectMake(20, 20, 50, 50)];
    view.backgroundColor = [AppTheme textColor];
    FWBadgeView *badgeView = [[FWBadgeView alloc] initWithBadgeStyle:FWBadgeStyleDot];
    [view fw_showBadgeView:badgeView badgeValue:nil];
    [self.view addSubview:view];
    
    view = [[UIView alloc] initWithFrame:CGRectMake(20, 90, 50, 50)];
    view.backgroundColor = [AppTheme textColor];
    badgeView = [[FWBadgeView alloc] initWithBadgeStyle:FWBadgeStyleSmall];
    [view fw_showBadgeView:badgeView badgeValue:@"9"];
    [self.view addSubview:view];
    
    view = [[UIView alloc] initWithFrame:CGRectMake(90, 90, 50, 50)];
    view.backgroundColor = [AppTheme textColor];
    badgeView = [[FWBadgeView alloc] initWithBadgeStyle:FWBadgeStyleSmall];
    [view fw_showBadgeView:badgeView badgeValue:@"99"];
    [self.view addSubview:view];
    
    view = [[UIView alloc] initWithFrame:CGRectMake(160, 90, 50, 50)];
    view.backgroundColor = [AppTheme textColor];
    badgeView = [[FWBadgeView alloc] initWithBadgeStyle:FWBadgeStyleSmall];
    [view fw_showBadgeView:badgeView badgeValue:@"99+"];
    [self.view addSubview:view];
    
    view = [[UIView alloc] initWithFrame:CGRectMake(20, 160, 50, 50)];
    view.backgroundColor = [AppTheme textColor];
    badgeView = [[FWBadgeView alloc] initWithBadgeStyle:FWBadgeStyleBig];
    [view fw_showBadgeView:badgeView badgeValue:@"9"];
    [self.view addSubview:view];
    
    view = [[UIView alloc] initWithFrame:CGRectMake(90, 160, 50, 50)];
    view.backgroundColor = [AppTheme textColor];
    badgeView = [[FWBadgeView alloc] initWithBadgeStyle:FWBadgeStyleBig];
    [view fw_showBadgeView:badgeView badgeValue:@"99"];
    [self.view addSubview:view];
    
    view = [[UIView alloc] initWithFrame:CGRectMake(160, 160, 50, 50)];
    view.backgroundColor = [AppTheme textColor];
    badgeView = [[FWBadgeView alloc] initWithBadgeStyle:FWBadgeStyleBig];
    [view fw_showBadgeView:badgeView badgeValue:@"99+"];
    [self.view addSubview:view];
}

#pragma mark - Action

- (BOOL)shouldPopController
{
    [self onClose];
    return NO;
}

- (void)onClose
{
    FWWeakifySelf();
    [self fw_showConfirmWithTitle:nil message:@"是否关闭" cancel:nil confirm:nil confirmBlock:^{
        FWStrongifySelf();
        [self fw_closeViewControllerAnimated:YES];
    }];
}

- (void)onClick:(UIBarButtonItem *)sender
{
    [sender fw_hideBadgeView];
}

@end

File: Example/FWFramework/Classes/Module/Test/Module/TestBadgeController.swift
//
//  TestBadgeController.swift
//  FWFramework_Example
//
//  Created by wuyong on 2022/9/21.
//  Copyright © 2022 CocoaPods. All rights reserved.
//

import FWFramework

class TestBadgeController: UIViewController, ViewControllerProtocol {
    
    func didInitialize() {
        hidesBottomBarWhenPushed = false
        app.extendedLayoutEdge = .bottom
    }
    
    func setupNavbar() {
        app.tabBarHidden = false
        app.setLeftBarItem(Icon.backImage, target: self, action: #selector(onClose))
        var badgeView = BadgeView(badgeStyle: .dot)
        navigationItem.leftBarButtonItem?.app.showBadgeView(badgeView)
        
        let rightItem = UIBarButtonItem.app.item(object: Icon.backImage, target: self, action: #selector(onClick(_:)))
        badgeView = BadgeView(badgeStyle: .small)
        rightItem.app.showBadgeView(badgeView, badgeValue: "1")
        
        let customView = UIButton(frame: CGRect(x: 0, y: 0, width: 20, height: 20))
        customView.backgroundColor = AppTheme.textColor
        let customItem = UIBarButtonItem.app.item(object: customView, target: self, action: #selector(onClick(_:)))
        badgeView = BadgeView(badgeStyle: .small)
        customItem.app.showBadgeView(badgeView, badgeValue: "1")
        navigationItem.rightBarButtonItems = [rightItem, customItem]
    }
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        
        var badgeView = BadgeView(badgeStyle: .dot)
        tabBarController?.tabBar.items?[0].app.showBadgeView(badgeView)
        badgeView = BadgeView(badgeStyle: .small)
        tabBarController?.tabBar.items?[1].app.showBadgeView(badgeView, badgeValue: "99")
    }
    
    override func viewWillDisappear(_ animated: Bool) {
        super.viewWillDisappear(animated)
        
        tabBarController?.tabBar.items?[0].app.hideBadgeView()
        tabBarController?.tabBar.items?[1].app.hideBadgeView()
    }
    
    func setupSubviews() {
        var view = UIView(frame: CGRect(x: 20, y: 20, width: 50, height: 50))
        view.backgroundColor = AppTheme.textColor
        var badgeView = BadgeView(badgeStyle: .dot)
        view.app.showBadgeView(badgeView)
        self.view.addSubview(view)
        
        view = UIView(frame: CGRect(x: 20, y: 90, width: 50, height: 50))
        view.backgroundColor = AppTheme.textColor
        badgeView = BadgeView(badgeStyle: .small)
        view.app.showBadgeView(badgeView, badgeValue: "9")
        self.view.addSubview(view)
        
        view = UIView(frame: CGRect(x: 90, y: 90, width: 50, height: 50))
        view.backgroundColor = AppTheme.textColor
        badgeView = BadgeView(badgeStyle: .small)
        view.app.showBadgeView(badgeView, badgeValue: "99")
        self.view.addSubview(view)
        
        view = UIView(frame: CGRect(x: 160, y: 90, width: 50, height: 50))
        view.backgroundColor = AppTheme.textColor
        badgeView = BadgeView(badgeStyle: .small)
        view.app.showBadgeView(badgeView, badgeValue: "99+")
        self.view.addSubview(view)
        
        view = UIView(frame: CGRect(x: 20, y: 160, width: 50, height: 50))
        view.backgroundColor = AppTheme.textColor
        badgeView = BadgeView(badgeStyle: .big)
        view.app.showBadgeView(badgeView, badgeValue: "9")
        self.view.addSubview(view)
        
        view = UIView(frame: CGRect(x: 90, y: 160, width: 50, height: 50))
        view.backgroundColor = AppTheme.textColor
        badgeView = BadgeView(badgeStyle: .big)
        view.app.showBadgeView(badgeView, badgeValue: "99")
        self.view.addSubview(view)
        
        view = UIView(frame: CGRect(x: 160, y: 160, width: 50, height: 50))
        view.backgroundColor = AppTheme.textColor
        badgeView = BadgeView(badgeStyle: .big)
        view.app.showBadgeView(badgeView, badgeValue: "99+")
        self.view.addSubview(view)
    }
    
    override var shouldPopController: Bool {
        onClose()
        return false
    }
    
    @objc func onClose() {
        app.showConfirm(title: nil, message: "是否关闭") { [weak self] in
            self?.app.close()
        }
    }
    
    @objc func onClick(_ sender: UIBarButtonItem) {
        sender.app.hideBadgeView()
    }
    
}
